sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java

```diff
@@ -483,6 +488,54 @@ public String toString() {
     return Read.newBuilder(parsePayloadUsingCoder(coder)).setCoder(coder).build();
   }
 
+  /**
+   * Returns a {@link PTransform} that continuously reads binary encoded protobuf messages for the
+   * type specified by {@code fullMessageName}.
+   *
+   * <p>This is primarily here for cases where the message type cannot be known at compile time. If
+   * it can be known, prefer {@link PubsubIO#readProtos(Class)}, as {@link DynamicMessage} tends to
+   * perform worse than concrete types.
+   *
+   * <p>Beam will infer a schema for the {@link DynamicMessage} schema. Note that some proto schema
+   * features are not supported by all sinks.
+   *
+   * @param domain The {@link ProtoDomain} that contains the target message and its dependencies.
+   * @param fullMessageName The full name of the message for lookup in {@code domain}.
+   */
+  @Experimental(Kind.SCHEMAS)
+  public static Read<DynamicMessage> readProtoDynamicMessage(
+      ProtoDomain domain, String fullMessageName) {
+    SerializableFunction<PubsubMessage, DynamicMessage> parser =
+        message -> {
+          try {
+            return DynamicMessage.parseFrom(
+                domain.getDescriptor(fullMessageName), message.getPayload());
+          } catch (InvalidProtocolBufferException e) {
+            throw new RuntimeException("Could not parse Pub/Sub message", e);
```

Comment:
Does PubsubIO have a dead-letter capability that will catch this error and output unparseable messages to some secondary output? Can users turn that capability on after the `Read` transform is returned by this method?

(I'm not opening up the code and reading, because I just want to get your review back to you and I'm sure you know the answer and will do the right thing)

Reply:
Currently, it isn't supported, and the other methods that parse binary content for Protos or Avro throw a RuntimeException like here. 

I do think it would be valuable to have something like `withDeadLetter` to configure the returned `Read` object, but I think that's for a separate PR/Jira.

Reply:
Opened BEAM-12474 for tracking effort on adding this. I'll probably get around to working on it myself soon.